### DISCUSSION TITLE
Is there a way to rerun a successfully completed Future?

### DISCUSSION THREAD

**Discussion by ghost** [dask/distributed#4804] [Q&A] [answered]
I see that `client.recreate_error_locally` already exists. But, I don't see a publicly exposed way to do the same thing for successfully completed futures, does this already exist?  This can be useful if a task completes successfully, but produces some unexpected results that you want to debug.

**Comment by fjetter** [comment]
afaik, there is no standard API to do this but there is enough to put a recipe together and/or contribute this as a feature. For non-erred tasks I would argue, the logic is almost identical to the replay exception case. What the `recreate_error_locally` call does is to fetch the task runspec (i.e. the function + arguments) and calculates all the dependencies of that task.

This happens here https://github.com/dask/distributed/blob/5e150aa6a3bd7decd38193a3d754497f8914d83c/distributed/recreate_exceptions.py#L125-L131

For a non-erred task, the only thing we'll need to change is the `cause_of_failure` (in `_get_futures_error`) which retrieves information from the scheduler. This could look like

```python
def get_runspec(dask_scheduler, key):
    ts = dask_scheduler.tasks[key]
    return {
        "task": ts.run_spec,
        "deps": [dts.key for dts in ts.dependencies]
    }
client.run_on_scheduler(get_runspec)
```
then continue with https://github.com/dask/distributed/blob/5e150aa6a3bd7decd38193a3d754497f8914d83c/distributed/recreate_exceptions.py#L86-L92

I haven't tested this, therefore no guarantees but that should be roughly how it could look like. If you want to get your hands dirty, feel free to open a PR!

In a sense this would be a "calculate all dependencies on the cluster but the final task on my machine"-feature

**Reply by ghost** [reply]
Hey, will take a look at this soon. Do you think it would make sense to consolidate this concept and the existing recreate_error_locally?

**Reply by ghost** [reply]
@fjetter - opened a PR with an initial concept here:
https://github.com/dask/distributed/pull/4813

**Comment by ghost** [comment, marked as answer]
This now exists with `client.recreate_task_locally`
